### DISCUSSION TITLE
"refusing to create treatment more than 1 hour in the future"

### DISCUSSION THREAD

**Discussion by rpm5099** [NightscoutFoundation/xDrip#3343] [Bug reports]
I will provide as much detail and logs as necessary, please let me know what is required.  Any help is greatly appreciated.  I have the same problem on 3+ android devices and at least 4+ different versions of xdrip.  This has been verified with the latest xdrip nightly build from 2/14/2014 and AAPS compiled from git master HEAD less than 7 days ago.

Main question up front:  is it possible to use xdrip (with follower), AAPS (with client) and nightscout server and have treatments/notes (regardless of what platform they are entered from) synchronized across all?

The xdrip error originates from here:
`app/src/main/java/com/eveningoutpost/dexdrip/models/Treatments.java:255:            JoH.static_toast_long("Refusing to create a treatement more than 1 hours in the future!");`

My daughter's phone  (android Nord n30) has xdrip connected to a dexcom G6 with xdrip and AAPS installed.  xdrip is the collector from the sensor.  My phone (galaxy s23 ultra) has xdrip and AAPSclient installed, xdrip as sync follower.  I have a nightscout server hosted locally which seems to work perfectly.  

I've read virtually all of the documentation, cloned the gitrepos and searched through source code, and tried every combination of every setting on xdrip, aaps, and nightscout server and am getting nowhere.

The main problem with xdrip seems to be that xdrip does not have the correct time or timezone and constantly shows an error that treatments cannot be entered in the future, regardless where they are being entered - Nightscout web, AAPS/AAPS client, xdrip follower.  My wife and I are unable to enter treatments (insulin/carbs/notes) and have them sync across xdrip - required for BG predictions based on COB/IOB.  Nightscout and AAPS seem to work fine.  I cannot move forward with an insulin pump with this problem, and I cannot coordinate information with my wife using dexcom's terrible apps, but I have spent so many hours trying to figure this out and getting nowhere that I'm about to give up and do that.

Xdrip follower sync does not consistently mirror treatments in both directions, and never goes from master to follower.  Neither xdrip will accept any treatment from the nightscout server.  Should you have nsclient sync setup on just the xdrip master or also the follower?  What features of nsclient are bidirectional?  Shouldn't the follower simply receive ALL information (treatments, bg, notes, calibration, profile, etc) directly from the xdrip master, and master receive updates for all except BG?  Should AAPS master be connected to the xdrip master only, or does it also need to be connect to nsclient?  How do they deconflict?  How are data entries prevented from duplication via repeating?  What acts as the source/destination for which information, or if it's configurable where is that in the documentation?  

What is the expected/possible dataflow?  Are there any diagrams indicating what is possible or correct?

- xdrip master to xdrip follower:
   - treatments?
   - bg - yes
   - notes?
- xdrip follower to xdrip master
  - treatments?
  - bg - no
  - notes?
- xdrip master to nightscout server
  - treatments, yes?
  - bg - YES
  - notes?
- nightscout server to xdrip master
  - treatments - hopefully?
  - bg - NO
  - notes - hopefully?
- xdrip follower to nightscout server
  - treatments - why not directly to xdrip master via sync?  (does not work)
  - bg - NO
  - notes - same as treatments
- nightscout server to xdrip follower
  - treatments/notes/bg - why not from xdrip master sync
 
- xdrip master to AAPS master
  - BG - YES
  - treatments - does not work
  - notes - does not work
- AAPS master to AAPS follower
 - ...

**Comment by PhilTU** [comment]
I haven't used this in anger, but at https://navid200.github.io/xDrip/docs/xDripSync.html you can see some information. 

An xdrip sync follower should be able to upload -  "By default*, the follower can upload treatments to the master, calibrate it, and even stop sensor on it. Disable Allow follower actions to remove that possibility."

"Accept follower actions" is an xDrip+ sync setting on the master.

There is an additional setting to "Disable all sync features" in the xDrip+ sync settings which would override the above.

For xDrip+ <> Nightscout there are again settings to enable Treatments download and upload on the master.
Apologies if you've checked them already.

For nightscout server to xdrip follower you could make the xDrip follower a Nightscout follower instead as an alternative ?

On the time front are all devices set to the same time zone and DST settings ? I had an unrelated problem with one device set to GMT+1 and another at GMT with DST enabled, both read the same at first glance but were interpreted differently.

I haven't read anything about "synchronisation" in the true sense - where one set of data appears identically across devices. Closest would be for the master to upload to Nightscout and others to follow NS so they're all on the same page.

**Reply by 808IUFan** [reply]
This issue just started for me and I only have one device using xDrip. I think we need an update.

**Reply by Navid200** [reply]
@808IUFan What exactly is the problem?

**Reply by 808IUFan** [reply]
No readings since that message started popping an hour ago. I wonder if maybe wherever the data is switched to daylight savings?

**Reply by 808IUFan** [reply]
![Screenshot_20250309-010311](https://github.com/user-attachments/assets/4e4cc429-ad65-4da2-b36f-10ac45a7379c)

**Reply by Navid200** [reply]
No way to know.  You have not provided any info.  
If you are a follower, are both master and follower time settings are synchronized to the internet?

**Reply by 808IUFan** [reply]
Did you see my messaage and the screenshot? When the "Refusing to create a treatement more than 1 hours" appeared my readings stopped from Carelink. Someone else posted a novel up there but most does not apply to me. I have one phone so I have no master and follower settings.

**Reply by Navid200** [reply]
What do you think CareLink is?
You are the one who needs help.  If you treat anyone who attempts to help you by questioning their approach, it will not be very helpful to yourself.
Good luck finding a solution.

**Comment by 808IUFan** [comment]
Carelink is where you get my data from for xDrip. Why do you always think I am not cooperative?? Your message "what is the problem" sounded sarcastic to me so I asked if you got my stated problem in the first message and the screenshot. 2 other people have reported the same thing. We get a message and no readings from our pump/carelink. I am up to date with versions, my settings have not changed. I see "refusing to create treatement (not spelled correct in the message) more than 1 hour in the future." Please let me know what you need to help. I have also attached the log file. 
[xDrip-log.txt](https://github.com/user-attachments/files/19149853/xDrip-log.txt)

**Reply by Navid200** [reply]
The time on your phone has changed I suppose because of daytime saving.  Have you updated the time on the pump accordingly?

**Comment by 808IUFan** [comment]
Yes I have. The times match exactly on my phone and pump. I suggested last night that could be the problem whereever this data is processed but both match for me. I did change the time on my pump before I went to bed but then I changed it back thinking that could have caused the issue. I waited until 2am to update the time on my pump.

**Reply by Navid200** [reply]
Change the xDrip hardware data source to disabled.  Wait 5 minutes.
Then, change it back to CareLink.  Does it work?

**Reply by PhilTU** [reply]
How is the glucose data entered, the log shows errors relating to blood test glucose values from this point on :- 

^ 2025-03-08 23:46:06 ^ BloodTest ^ Timestamp is > 10 minutes in the future! Something is wrong: 2025-03-09 01:35:00

**Reply by 808IUFan** [reply]
I have to calibrate my pump twice a day from a meter but times match on that also because it automatically syncs with my pump. Disabled collections for more than 5 minutes and still the same error and no readings.

**Reply by Navid200** [reply]
Have you restarted the phone?

**Reply by PhilTU** [reply]
> I have to calibrate my pump twice a day from a meter but times match on that

ok, does xDrip receive them as treatments from the meter ?

**Reply by 808IUFan** [reply]
Phone has been restarted and the app does not receive treatments from the meter just blood glucose. Treatments come from the pump and carelink. I just noticed that all of the other data, like units on board, insulin reserve in the pump and the battery status is correct. Just no readings. Even the trend line is there. Still have Signal missed 597 minutes ago and the reading is that last reading from last night with the line through it. I have a feeling I am going to have to uninstall and reinstall.

**Reply by Navid200** [reply]
What do you see in the calibration data table?
https://navid200.github.io/xDrip/docs/Datatables.html

**Reply by 808IUFan** [reply]
Calibration table is blank.

**Reply by Navid200** [reply]
Does your meter have a queue that it is trying to submit to xDrip that you can clear so that it would stop sending it?

**Reply by 808IUFan** [reply]
No I only used the meter twice a day and it sends to the pump automatically. No bluetooth transmissions have failed either.

**Reply by Navid200** [reply]
Oh I see it is coming from the pump.

The log that Phil mentioned shows a blood test that is being submitted to xDrip that it finds more than 10 minutes in the future and rejects.  It is possible that the pump keeps sending it and as long as it is not accepted, it is holding off on sending anything else.
I am just speculating here.

**Reply by 808IUFan** [reply]
Not sure about that because I have not seen any errors. I went ahead and uninstall, deleted the database loaded to my phone and reinstalled. Same message is there.  It pops when the app tries to poll for a reading.

**Reply by Navid200** [reply]
How about CareLink?  Can you log into it directly and see if it keeps submitting a blood test?

**Comment by 808IUFan** [comment]
For patients we cant see that and this is SUPER embarrassing. I just found the issue. When I changed the time last night I didnt verify AM and PM so the numbers were correct but it was off by 12 hours. I apologize profusely.

**Reply by PhilTU** [reply]
Aha, I was thinking it might be the *date* that was wrong after you changed time over midnight, well spotted.

**Comment by 808IUFan** [comment]
Yeah changing the time on Medtronic pumps is not setup well. Sometimes you have to hit ENTER to change things and sometimes that moves to the next item. I found it when I opened the Carelink Website to test uploading data and it said my time was off by 12 hours. Thank you both for your help again.

**Reply by jamorham** [reply]
Good to know this doesn't appear to be an xDrip problem and that xDrip alerted you to the incorrect time issue. Shame the pump doesn't automatically update its time.

**Reply by 808IUFan** [reply]
It wasn't xDrip. It was Carelink. I went to see if that worked and got an immediate message that my AM PM were wrong and time was off by 12 hours.
